### PR TITLE
Initial support for project references

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace Microsoft.CodeAnalysis.Testing
+{
+    public abstract class ProjectState
+    {
+        private readonly string _name;
+        private readonly string _defaultPrefix;
+        private readonly string _defaultExtension;
+
+        protected ProjectState(string name, string defaultPrefix, string defaultExtension)
```
Comment (jmarolf) [inline src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs]: nit: Perhaps call this `ProjectData``?

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            _name = name;
+            _defaultPrefix = defaultPrefix;
+            _defaultExtension = defaultExtension;
+
+            Sources = new SourceFileList(defaultPrefix, defaultExtension);
+        }
+
+        public string Name { get; set; }
```
Comment (jmarolf) [inline src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs]: Its not obvious to me why you can set this value after creating the object

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs (hunk 1 above)
Comment (jmarolf) [reply src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs]: nevermind this is to match `SolutionState`.

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs (hunk 2 above)
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs]: I'll make it read only

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs (hunk 1 above)
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ProjectState.cs]: Yes. I have a good feeling this API will change forms (e.g. to support diagnostics/markup/Fix All in Solution testing), but I'm hoping it's a decent start to unblock some testing for @mavasani.

**Pull request merged**